File: aoj/alds1/4_search/b_binary_search.cpp
#include <iostream>

using namespace std;

int binary_search(int n, int* S, int v) {
  int s = 0, t = n;
  while (s < t) {
    int j = (s + t) / 2;
    if (v == S[j]) {
      return 1;
    } else if (v < S[j]) {
      t = j;
    } else {
      s = j+1;
    }
  }
  return 0;
}



int main() {
  int n,q,v,cnt;
  cin >> n;
  int S[n];
  for (int i = 0; i < n; i++) {
    cin >> S[i];
  }

  cin >> q;
  cnt = 0;
  for (int i = 0; i < q; i++) {
    cin >> v;
    cnt += binary_search(n, S, v);
  }

  cout << cnt << endl;
  return 0;
}

File: aoj/alds1/4_search/b_binary_search.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strconv"
)

func binarySearch(list []int, target int, l int, u int) int {
	if u - l <= 1 {
		if list[u] == target {
			return u
		} else {
			return -1
		}
	}

	m := (u + l) / 2
	if target <= list[m] {
		return binarySearch(list, target, l, m)
	} else {
		return binarySearch(list, target, m, u)
	}
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	
	sc.Scan()
	n, _ := strconv.Atoi(sc.Text())

	s := make([]int, n)
	for i := 0; i < n; i++ {
		sc.Scan()
		s[i], _ = strconv.Atoi(sc.Text())
	}

	sc.Scan()
	q, _ := strconv.Atoi(sc.Text())
	
	cnt := 0
	for i := 0; i < q; i++ {
		sc.Scan()
		target, _ := strconv.Atoi(sc.Text())
		
		result := binarySearch(s, target, -1, n)
		if result >= 0 {
			cnt++
		}
	}
	fmt.Println(cnt)
}
